gpu4pyscf/fci/direct_spin1.py

```diff
@@ -0,0 +1,191 @@
+    int stra = stra0 + ty;
+    int nab = na * TILE;
+    int ab_id = stra * TILE + tx;
+    __shared__ unsigned short _addra[TILE*TILE];
+    __shared__ unsigned short _addrb[TILE*TILE];
+    __shared__ char _signa[TILE*TILE];
+    __shared__ char _signb[TILE*TILE];
+    int sign, str1, j0, j;
+    int dj = TILE;
+    double val = 0.;
+
+    for (j0 = 0; j0 < nnorb; j0+=TILE) {
+        _addra[ty*TILE+tx] = addra[(j0+ty)*na+stra0+tx];
+        _addrb[ty*TILE+tx] = addrb[(j0+ty)*nb+strb0+tx];
+        _signa[ty*TILE+tx] = signa[(j0+ty)*na+stra0+tx];
+        _signb[ty*TILE+tx] = signb[(j0+ty)*nb+strb0+tx];
+        if (j0 + TILE > nnorb) {
+            dj = nnorb - j0;
+        }
+        __syncthreads();
+        if (stra < na && strb < nb) {
+            for (j = 0; j < dj; j++) {
+                sign = _signa[j*TILE+ty];
+                str1 = _addra[j*TILE+ty];
+                if (sign != 0) {
+                    val += sign * t1[(j0+j)*nab + (str1*TILE+tx)];
+                }
+
+                sign = _signb[j*TILE+tx];
+                str1 = _addrb[j*TILE+tx];
+                if (sign != 0) {
+                    out[stra*nb+str1] += sign * t1[(j0+j)*nab + ab_id];
+                }
+            }
+        }
+        __syncthreads();
+    }
+    out[stra*nb+strb] += val;
+}
+}'''
+
+_contract_2e_spin1 = cp.RawModule(code=code)
+_build_t1 = _contract_2e_spin1.get_function('_build_t1')
+_gather = _contract_2e_spin1.get_function('_gather')
+
+def _link_index_to_addrs(link_index, nnorb):
+    na, nov = link_index.shape[:2]
+    ia = link_index[:,:,0].T
+    addr = np.zeros((nnorb, na), dtype=np.uint16)
+    sign = np.zeros((nnorb, na), dtype=np.int8)
+    #:for j in range(nov):
+    #:    for a in range(na):
+    #:        addr[ia[a,j],a] = link_index[a,j,2]
+    #:        sign[ia[a,j],a] = link_index[a,j,3]
+    idx = np.arange(na)
+    addr[ia,idx] = link_index[:,:,2].T
+    sign[ia,idx] = link_index[:,:,3].T
+    # Add paddings to avoid illegal address in kernel
+    _addr = cp.empty((nnorb+TILE-1, na), dtype=np.uint16)[:nnorb]
+    _sign = cp.empty((nnorb+TILE-1, na), dtype=np.int8)[:nnorb]
+    _addr.set(addr)
+    _sign.set(sign)
+    return _addr, _sign
+
+def contract_2e(eri, ci0, norb, nelec, link_index):
+    ci0 = cp.asarray(ci0)
+    out = cp.zeros_like(ci0)
+    na, nb = ci0.shape
+    nnorb = norb * (norb + 1) // 2
+    assert eri.shape == (nnorb, nnorb)
+    eri = cp.asarray(eri)
+    link_indexa, link_indexb = link_index
+    addra, signa = _link_index_to_addrs(link_indexa, nnorb)
+    if link_indexa is link_indexb:
+        addrb, signb = addra, signa
+    else:
+        addrb, signb = _link_index_to_addrs(link_indexb, nnorb)
+
+    threads = (TILE, TILE)
+    blocks = (1, (na+TILE-1)//TILE)
+    rest_args = (na, nb, nnorb, addra, addrb, signa, signb)
+    t1 = cp.empty((nnorb, na*TILE))
+    gt1 = cp.empty((nnorb, na*TILE))
+    if 0:
+        # Pipeline the three operations. Seems no performance improvement
```

Comment:
Is it because _gather is running in the default stream? The default stream has to run synchronously w.r.t. other streams.

Reply:
I tried that, no help. I guess the memory bandwidth is probably saturated. I was hoping to overlap the dot operation and the other two data moving functions. However, the dot operation also occupies most of the memory bandwidth, since the product is between a tall skinny matrix and a small square matrix.